course1/course_en.md

```diff
@@ -0,0 +1,199 @@
+</style>
+
+# Course Outline
+
+<div class="columns">
+<div>
+
+| # | Topic |
+|-:|:-:|
+| 1 | Introduction & Program Design |
+| 2 | Dev Environments & Expressions |
+| 3 | Functions, Lists & Recursions |
+| 4 | Tuples, Structs & Enums |
+| 5 | Trees |
+| 6 | Generics & Higher-Order Functions |
+| 7 | Imperative Programming |
+| 8 | Queues |
+
+</div>
+<div>
+
+| # | Topic |
+|-:|:-:|
+| 9 | Traits |
+| 10 | Hash Tables & Closures |
+| 11 | Case Study: Parsers |
+| 12 | Case Study: Autodiff |
+| 13 | Case Study: Neural Networks |
+
+- All materials are available online
+- Forum: [discuss.moonbitlang.com](https://discuss.moonbitlang.com/)
+
+</div>
+</div>
+
+---
+
+# Program Design
+
+---
+
+# Basic Workflow of Program Design
+
+**Program design** is the process of converting informal specifications into a program.
+
+It is highly recommended to adopt a **test-driven development** (TDD) workflow, _viz._,
+
+1. **Understand the problem**: What variables are involved and how are they related?
+2. **Define the interfaces**: How should the program interact with the environment?
+3. **Write the test cases**: What is the expected behavior for normal inputs? What can be the potential abnormal inputs?
+4. **Implement the program**: Usually, it is necessary to break down the problem into smaller subproblems and repeat the above process for each of them.
+
+---
+
+# An Example of Program Design
+
+> There are `num_bottles` water bottles that are initially full of water. You can exchange `num_exchange` empty water bottles from the market with one full water bottle.
+> The operation of drinking a full water bottle turns it into an empty bottle.
+> Given the two integers `num_bottles` and `num_exchange`, return _the **maximum** number of water bottles you can drink_.
+> <span style="float:right;">Source: [LeetCode 1518](https://leetcode.com/problems/water-bottles/description/)</span>
+
+---
+
+# Step 1: Understand the Problem
+
+- What variables are involved?
+    - $N_\mathrm{bottles}$: The number of _full_ water **bottles** that we currently have.
+    - $N_\mathrm{drunk}$: The number of _empty_ water bottles that we have **drunk**.
+- How are they related?
+    - The initial value for $N_\mathrm{bottles}$ is given by the input.
+    - When $N_\mathrm{bottles} \ge$ `num_exchange`, we can drink `num_exchange` of them, exchange them for one full water bottle, and then repeat the process.
+    - When $N_\mathrm{bottles} <$ `num_exchange`, we have to drink them up and quit.
+- What do we want to calculate?
+    - Given `num_bottles` and `num_exchange`, calculate the maximum number of water bottles we can drink.
+
+---
+
+# Step 2: Define the Interfaces
+
+> Given the two integers `num_bottles` and `num_exchange`, return _the **maximum** number of water bottles you can drink_.
+
+```moonbit
+fn num_water_bottles(num_bottles: Int, num_exchange: Int) -> Int {
+    abort("To be done")
+}
+```
+
+---
+
+# Step 3: Write the Test Cases
+
+```moonbit
```

Comment:
We might introduce MoonBit core very soon, so we may use the `@assertion.assert_true` directly. The integration will be done this week, so we'll update this part next week.

Reply:
```moonbit
test {
  @assertion.assert_eq(num_water_bottles(9, 3), 13)? // 9 + 3 + 1 = 13
  @assertion.assert_eq(num_water_bottles(15, 4), 19)?
}
```